Fix #6792: Adding Exploration Titles to Collection Landing Page

Showtim3 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #6792 
**Actually its a new feature addition.** So the thing is I added the titles above the explorations on the collection homepage.
**Small Size Devices:** 
![mobile-view](https://user-images.githubusercontent.com/30312043/59162458-226f4a80-8b0f-11e9-98ca-9a425b91bc1e.png)
**Desktop Size Devices:**
![desktop-view](https://user-images.githubusercontent.com/30312043/59162459-226f4a80-8b0f-11e9-8f43-fb4e6d219ac6.png)

I'd like some feedback on the changes i should do.
## Checklist
- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [X] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [X] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [X] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Showtim3 committed:
Extended hover on about menu when any li child is hovered

Showtim3 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into extend-hover-effect-on-aboutMenu

Showtim3 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into extend-hover-effect-on-aboutMenu

Showtim3 committed:
updated with develop branch

Showtim3 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into extend-hover-effect-on-aboutMenu

Showtim3 committed:
reverted package-lock.json

Showtim3 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into titles-to-explorations

Showtim3 committed:
Adding titles to explorations on collection homepage

Showtim3 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into titles-to-explorations

Showtim3 commented:
@import-keshav @dchen97

import-keshav reviewed (changes requested):
Thanks, @Showtim3 for the PR, and I appreciate your work (for finding which file to debug) :) :)

While there is no comment in code from my side, but I would say can you implement the frontend part as mentioned in the issue? As the images in the issue are more user-friendly than this.

Thanks

Showtim3 commented:
@import-keshav I totally agree with you. This issuse isn't complete yet, I'll work on the UI part now.
Thanks!

Showtim3 committed:
Fixed UI of titles to explorations

Showtim3 committed:
Fixed linting

Showtim3 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into titles-to-explorations

Showtim3 commented:
@import-keshav @kevinlee12  @nithusha21 
**Desktop View :**
![new](https://user-images.githubusercontent.com/30312043/59550737-ee22e080-8f8b-11e9-9126-84df674d1348.png)
**Mobile and tablet View:**
![mobNew](https://user-images.githubusercontent.com/30312043/59550738-ee22e080-8f8b-11e9-8455-9748ed0a365b.png)

Its looking pretty decent,I guess everything is fine now.

import-keshav commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.html:
```diff
@@ -47,12 +47,18 @@ <h2 ng-if="$ctrl.collectionPlaythrough.hasFinishedCollection()" class="oppia-pag
+
+        <div ng-if="[$index]%2===0" ng-style="{width:'160',position:'absolute',color:'#006553','font-size':'20px',bottom:'85%'}">{{node._explorationSummaryObject.title}}</div>
+        <div ng-if="[$index+1]%2===0&&[$index+1]%4!==0" ng-style="{width:'160',position:'absolute',color:'#006553',bottom:'85%','font-size':'20px', left:'20px'}">{{node._explorationSummaryObject.title}}</div>
+        <div ng-if="[$index+1]%4===0" ng-style="{width:'160px',position:'absolute',color:'#006553','font-size':'20px',bottom:'85%',left:'-35px'}">{{node._explorationSummaryObject.title}}</div>
```
Try to do all the ``` ng-if ``` functionality, in a function. Don't do all this operation in HTML. Do in Js file.
Like ``` ng-if = function( ) ``` .

import-keshav reviewed (changes requested):
Thanks, @Showtim3 for the PR. Its LGTM :)
Also, just to check, did you manually test with the longest possible state name?
And left one comment PTAL
Thanks

import-keshav commented:
@Showtim3 PTAL on failing Travis Tests.

Showtim3 commented:
@import-keshav hey this is when all the cards are of max chars i.e 40. Would it be better if i shifted the whole card component a bit down? so that the topmost card doesn't float around the box or is it fine this way?
![image](https://user-images.githubusercontent.com/30312043/59554881-a4f08200-8fc7-11e9-9ffa-79d95789ccfc.png)

import-keshav commented:
@Showtim3 its fine to shift whole card component a bit down. But please ensure that this doesn't break any other components.
Thanks

Showtim3 commented:
@import-keshav they look very pretty now and i made the function change you requested, and its definitely better this way, more cleaner.
**Desktop View:**
**![new-desktop-fixed](https://user-images.githubusercontent.com/30312043/59555490-ad00ef80-8fd0-11e9-82cf-0ca068e29554.png)**
**Mobile View:**
![mobile-new-fixed](https://user-images.githubusercontent.com/30312043/59555491-ad00ef80-8fd0-11e9-9045-9b66f05cba5d.png)

Showtim3 committed:
Fixed overflowing of titles outside the card component

Showtim3 committed:
Fixed linting

Showtim3 committed:
fixed linting-2

Showtim3 committed:
Fixed linting...

Showtim3 committed:
Fixed some more linting

Showtim3 committed:
Fixing Linting...

import-keshav reviewed (changes requested):
@Showtim3. Frontend View is now LGTM. Thanks
Just one comment from my side, otherwise PR is LGTM.
Thanks

import-keshav commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.ts:
```diff
@@ -250,6 +250,30 @@ oppia.directive('collectionPlayerPage', ['UrlInterpolationService',
+                'font-size': '20px',
+                bottom: '70%',
+                left: '-35px' };
+            }
```
Try to make it more generalized.
Like create one constant ``` exploration_title_css_dict ``` (doesn't know the naming convention).
Whose default value be
```
{
  width: '160px',
  position: 'absolute',
  color: '#006553',
  font-size: '20px',
}
```
And now add remaing key value pairs, according to if else condtion.

Showtim3 replied:
@import-keshav hey buddy thanks for being so patient with me. I made the change you suggested and now I am begining to understand why opensource is encouraged so much. :D

import-keshav replied:
Thanks, @Showtim3
And welcome to opensource :)

Showtim3 committed:
Made the function more generic

import-keshav reviewed (approved):
Thanks, @Showtim3 PR is LGTM!

import-keshav commented:
@kevinlee12 PTAL as code owner.
PTAL @nithusha21

kevinlee12 commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.ts:
```diff
@@ -250,6 +250,24 @@ oppia.directive('collectionPlayerPage', ['UrlInterpolationService',
               ctrl.collectionId);
           };
 
+          ctrl.getExplorationTitleStyle = function(index) {
```
Instead of generating the style here, is it possible to just have this function manipulate the `left` attribute instead? The reason is that everything else is fixed, so there's not really a reason to do this.

Showtim3 replied:
Done!

kevinlee12 commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.ts:
```diff
@@ -250,6 +250,24 @@ oppia.directive('collectionPlayerPage', ['UrlInterpolationService',
           };
 
+          ctrl.getExplorationTitleStyle = function(index) {
+            var explorationTitleStyle = {
```
This is missing a font attribute, the mock was using Capriola for the titles.

Showtim3 replied:
Done!

kevinlee12 commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.ts:
```diff
@@ -250,6 +250,24 @@ oppia.directive('collectionPlayerPage', ['UrlInterpolationService',
+              width: '160px',
+              position: 'absolute',
+              color: '#006553',
+              'font-size': '20px',
```
I would suggest making this smaller - 18px, since the titles look quite big on your screenshots.

Showtim3 replied:
Done!
@kevinlee12

kevinlee12 reviewed (changes requested):
Hi @Showtim3, I've left some comments, ptal!

Showtim3 committed:
Added font-family, changed font-size to 18 px and changed the function implementation

Showtim3 committed:
Fixed Linting

Showtim3 committed:
Removed unneccessary code

Showtim3 commented:
hey @kevinlee12 Any other suggestions?

kevinlee12 commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.html:
```diff
@@ -47,12 +47,25 @@ <h2 ng-if="$ctrl.collectionPlaythrough.hasFinishedCollection()" class="oppia-pag
          ng-href="<[$ctrl.getExplorationUrl(node.getExplorationId())]>"
          ng-style="{position: 'absolute', left: '<[$ctrl.pathIconParameters[$index].left]>', top: '<[$ctrl.pathIconParameters[$index].top]>'}">
+
+        <div ng-style="{width: '160px',
```
Instead of doing this, go for something like this file: https://github.com/oppia/oppia/blob/37a43ca249ffd2b60bf98f791995048ce0ec5269/core/templates/dev/head/components/summary_tile/exploration_summary_tile_directive.html. Specifically, move this into a `<style>` tag.

Then handle the left attribute like how we handle it in the library page: https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/library-page/library-page.directive.html

kevinlee12 commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.html:
```diff
@@ -47,12 +47,25 @@ <h2 ng-if="$ctrl.collectionPlaythrough.hasFinishedCollection()" class="oppia-pag
              xmlns:xlink="http://www.w3.org/1999/xlink"
-             version="1.1">
+             version="1.1"
+             ng-style="{position:'relative',top:'20px'}">
```
since these attributes are static, why can't we use the style tag directly.

Showtim3 replied:
I thought of letting angular take care of the styling, anyways yup it's not necessary. I'll put that up in the style tag.

kevinlee12 reviewed (changes requested):
Hi @Showtim3, looks like there was a bit of misunderstanding on what I meant by moving the style out, I've left an example (hope that helps!). As for UI, the text looks a bit too high and the text in the second exploration is hard to see since it overlaps with the path: 
![image](https://user-images.githubusercontent.com/5758256/59730034-b7500300-91f5-11e9-989c-48d2237f92a3.png)

Showtim3 commented:
> Hi @Showtim3, looks like there was a bit of misunderstanding on what I meant by moving the style out, I've left an example (hope that helps!). As for UI, the text looks a bit too high and the text in the second exploration is hard to see since it overlaps with the path:
> ![image](https://user-images.githubusercontent.com/5758256/59730034-b7500300-91f5-11e9-989c-48d2237f92a3.png)
Okay so how about we shift the righmost titles a bit more to the right? or maybe we can make the whole card component (the background one) a bit wider and then move the righmost headings a few pixels more to the right?
@kevinlee12

Showtim3 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into titles-to-explorations

Showtim3 committed:
Reduced bottom, changes ng-style to style

kevinlee12 commented:
I'm in favor of moving the rightmost titles a bit more to the right. It might also be good to move the leftmost titles a bit more to the left too.

Showtim3 committed:
Shifted increased left value of the leftmost titles

Showtim3 commented:
@kevinlee12 Hey, I reduced the bottom percentage a bit, changed the ng-style tags to style tags, adjusted the positions of both rightmost and leftmost titles. Do take a look and let me know if there's some change to be made.

kevinlee12 commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.html:
```diff
@@ -104,23 +111,15 @@ <h2 ng-if="$ctrl.collectionPlaythrough.hasFinishedCollection()" class="oppia-pag
       <div class="mobile-path-segment" ng-repeat="node in $ctrl.collection.getCollectionNodes()" id="mobile-path-anchor-<[$index]>">
-        <a href="" style="position: absolute; left: 50%; transform: translate(-50%, 195px); z-index: 1;">
+        <a href="" style="position: absolute; left: 50%; transform: translate(-50%, 195px); z-index: 1;text-align: center">
+          <div ng-style="{color:'#006553','font-size':'20px'}">{{node._explorationSummaryObject.title}}</div>
```
Convert this to `style` since there isn't anything here that is dynamic.

Showtim3 replied:
Yup Sure

Showtim3 replied:
Done!

kevinlee12 commented on core/templates/dev/head/pages/collection-player-page/collection-player-page.directive.html:
```diff
@@ -47,12 +47,19 @@ <h2 ng-if="$ctrl.collectionPlaythrough.hasFinishedCollection()" class="oppia-pag
          ng-href="<[$ctrl.getExplorationUrl(node.getExplorationId())]>"
          ng-style="{position: 'absolute', left: '<[$ctrl.pathIconParameters[$index].left]>', top: '<[$ctrl.pathIconParameters[$index].top]>'}">
+
+        <div style="width: 160px;position: absolute;left: <[$ctrl.getExplorationTitlePosition($index)]>;color: #006553;font-family : Capriola, Roboto, Arial,sans-serif;font-size: 18px;bottom: 65%;
```
I'm going to backtrack on this feedback, can you make it so that it looks more like line 49?

Showtim3 replied:
You mean one attribute per line? like width: 160px;
position: absolute? 
Or you mean like the single quotes ? like width : '160px'

kevinlee12 replied:
use `ng-style` instead of `style` for the attribute, everything else can remain as-is.

Showtim3 replied:
Done!

kevinlee12 reviewed (changes requested):
UI-wise, it looks good! Just a couple more nits, ptal!

Showtim3 committed:
Changed ng-style to style in tab vice and vice versa in desktop view

Showtim3 commented:
@kevinlee12 hey kevin, I pushed the changes that you asked, Do take a look. Thanks!

kevinlee12 reviewed (approved):
lgtm, thanks @Showtim3!

Showtim3 commented:
@kevinlee12 haha finally, Btw thank you so much for your guidance and patience despite me being stupuid again and again.

kevinlee12 commented:
my pleasure, you were great! :smile:
